lib/junit.rb

```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+require "rexml/document"
+require "rexml/xmldecl"
+require "rexml/cdata"
+
+module Homebrew
+  class Junit
+    BYTES_IN_1_MEGABYTE = 1024*1024
+    MAX_STEP_OUTPUT_SIZE = (BYTES_IN_1_MEGABYTE - (200*1024)).freeze # margin of safety
+
+    def initialize(tests)
+      @tests = tests
+    end
+
+    def build(filters: nil)
+      filters ||= []
+
+      @xml_document = REXML::Document.new
+      @xml_document << REXML::XMLDecl.new
+      testsuites = @xml_document.add_element "testsuites"
+
+      @tests.each do |test|
+        testsuite = testsuites.add_element "testsuite"
+        testsuite.add_attribute "name", "brew-test-bot.#{Utils::Bottles.tag}"
+        testsuite.add_attribute "tests", test.steps.count(&:passed?)
+        testsuite.add_attribute "failures", test.steps.count(&:failed?)
+        testsuite.add_attribute "timestamp", test.steps.first.start_time.iso8601
+
+        test.steps.each do |step|
+          next unless filters.any? { |filter| step.command_short.start_with? filter }
+
+          testcase = testsuite.add_element "testcase"
+          testcase.add_attribute "name", step.command_short
+          testcase.add_attribute "status", step.status
+          testcase.add_attribute "time", step.time
+          testcase.add_attribute "timestamp", step.start_time.iso8601
+
+          next unless step.output?
+
+          output = sanitize_output_for_xml(step.output)
+          cdata = REXML::CData.new output
+
+          if step.passed?
+            elem = testcase.add_element "system-out"
+          else
+            elem = testcase.add_element "failure"
+            elem.add_attribute "message",
+                               "#{step.status}: #{step.command.join(" ")}"
+          end
```

Comment:
I think we can probably strip the command output. It was pretty useful for Jenkins but it generates huge XML files that we don't really need/use now.

@jasonrudolph any ideas which of these fields BuildPulse requires/will use/will be useful?

Reply:
> any ideas which of these fields BuildPulse requires/will use/will be useful?

Let's look at an example XML file from rspec_junit_formatter for reference:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuite name="rspec" tests="2" skipped="0" failures="1" errors="0" time="0.015292" timestamp="2021-06-25T07:48:42-04:00" hostname="j19">
<properties><property name="seed" value="55788"/></properties>
<testcase classname="spec.example_0_spec" name="some-describe some-it" file="./spec/example_0_spec.rb" time="0.014212"><failure message="
expected: false
     got: true

(compared using ==)

Diff:
@@ -1,2 +1,2 @@
-false
+true
" type="RSpec::Expectations::ExpectationNotMetError">Failure/Error: expect(true).to eq false

  expected: false
       got: true

  (compared using ==)

  Diff:
  @@ -1,2 +1,2 @@
  -false
  +true

./spec/example_0_spec.rb:3:in `block (2 levels) in &lt;top (required)&gt;&apos;</failure></testcase>
<testcase classname="spec.example_spec" name="some-other-describe some-other-it" file="./spec/example_spec.rb" time="0.000072"></testcase>
</testsuite>
```

BuildPulse requires:
- [testsuite.name](https://github.com/Homebrew/homebrew-test-bot/blob/db8b7b2c2005438a45d0108d2cc6c6bb62e08057/lib/junit.rb#L25) (which is "rpsec" in the xample above)
- [testcase.name](https://github.com/Homebrew/homebrew-test-bot/blob/db8b7b2c2005438a45d0108d2cc6c6bb62e08057/lib/junit.rb#L34) (which is "some-describe some-it" in the first testcase above)
- [testcase.time](https://github.com/Homebrew/homebrew-test-bot/blob/db8b7b2c2005438a45d0108d2cc6c6bb62e08057/lib/junit.rb#L36) (which is "0.014212" in the first testcase above)
- For a failure, the testcase node needs a child [failure](https://github.com/Homebrew/homebrew-test-bot/blob/db8b7b2c2005438a45d0108d2cc6c6bb62e08057/lib/junit.rb#L47) node (shown in the first testcase above)
- For an error, the testcase node needs a child `error` node. It looks like the current code is [classifying any non-passing testcase as a failure](https://github.com/Homebrew/homebrew-test-bot/blob/db8b7b2c2005438a45d0108d2cc6c6bb62e08057/lib/junit.rb#L47), as opposed to distinguishing between errors and failures. That's totally fine from BuildPulse's perspective. 😅
- For failures and errors, BuildPulse accepts either a `message` attribute and/or `CDATA` to [describe the failure/error](https://twitter.com/jasonrudolph/status/1366447143783960579). (In the first testcase example above, the `message` and the CDATA are the same. I'm not yet sure why rspec_junit_formatter duplicates that info.)

If any of the following items are present, BuildPulse will use them:
- testcase.classname (which is "spec.example_0_spec" in the first testcase above)
- testcase.file (which is "./spec/example_0_spec.rb" in the first testcase above) -- This allows the BuildPulse UI/API to provide a link directly to the file that contains the flaky testcase.

BuildPulse ignores child nodes (like [system-out](https://github.com/Homebrew/homebrew-test-bot/blob/db8b7b2c2005438a45d0108d2cc6c6bb62e08057/lib/junit.rb#L45)) for passing tests.

Does that give you the info you're looking for?

Reply:
Thanks for the information @jasonrudolph! I'll double-check that we have all the required fields. 
For some PRs (like `rust` update for example) we could have quite a lot of failures with long `system-out`s. Are there any limits for the resulting file size? We could split a single JUnit file into several (by `testsute.name` for example), does BuildPulse support it?

Reply:
> Does that give you the info you're looking for?

Yes, thanks @jasonrudolph!

@bayandin I think we can just populate the stuff that will be used by BuildPulse?

Reply:
> @bayandin I think we can just populate the stuff that will be used by BuildPulse?

Yep, but it still requires error message:

> For failures and errors, BuildPulse accepts either a message attribute and/or CDATA to describe the failure/error. (In the first testcase example above, the message and the CDATA are the same. I'm not yet sure why rspec_junit_formatter duplicates that info.)

Or you suggest using some short cooked up error message instead of command output?

Reply:
@bayandin if we're only doing `audit` and `test` for now: I think that output length should always be fine? For `install` (if we add so now or in future) we probably just set a custom string rather than providing the entire install output?

Reply:
> For some PRs (like `rust` update for example) we could have quite a lot of failures with long `system-out`s. Are there any limits for the resulting file size?

Yes. BuildPulse currently rejects any XML files that are larger than 100 megabytes. 

> We could split a single JUnit file into several (by `testsute.name` for example), does BuildPulse support it?

Yes. You can have as many XML files as you like. :-)